### ISSUE TITLE
[HOLD for payment 2022-02-01]  File name isn't fully visible in Preview page

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#6873]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Lunch the app
2. Log in with any account
3. Select any user and tap on Plus button
4. Tap add attachment
5. Select from Galary

## Expected Result:
File name fully visible in Preview page

## Actual Result:
File name isn't fully visible in Preview page.

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.1.23.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

![attachment copy](https://user-images.githubusercontent.com/93399543/147019803-6dccd5bf-bee0-442b-8906-49a48e907853.png)

**Issue found by executing PR:** https://github.com/Expensify/App/pull/6850

**Expensify/Expensify Issue URL:**

**Issue reported by:**

**Slack conversation:**

**Comment by MelvinBot** [comment]
Triggered auto assignment to @marcaaron (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by marcaaron** [comment]
Not sure if this is a deploy blocker since it seems fairly minor and don't exist on production because the feature is not implemented there and not because of a regression.
 
@mananjadhav will maybe want to fix this as a follow up to https://github.com/Expensify/App/pull/6850

**Comment by mananjadhav** [comment]
@marcaaron I can add a fix for it. 

@kbecciv I couldn’t understand the issue. When you say it isn’t fully visible ? Is it getting trimmed or you mean the color visibility?

**Comment by kbecciv** [comment]
@mananjadhav Yes, the file name is trimmed only, no issue with a color.

**Comment by mananjadhav** [comment]
@marcaaron @shawnborton Can you help me get the intended behavior here? If the file name is too long this is how it looks.
![image](https://user-images.githubusercontent.com/3069065/147152128-a7eae7f1-8627-4819-b346-a127e6c01799.png)

Are we looking to trim it? At the moment I am reusing the subtitle of `Header` component. Also, are we still considering it a DeployBlocker?

**Comment by marcaaron** [comment]
No, not a deploy blocker.

It sounds like you are saying that anything with a "long subtitle" and using the `Header` component would cause this behavior and in that case probably does not need to be fixed before we can close out the linked issue.

As for a solution, probably would be best to loop in @Expensify/design. The two options seem to be:

- extend the height of the header to accommodate the large text
- use ellipsis to truncate the text

**Comment by MelvinBot** [comment]
Triggered auto assignment to @mallenexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by shawnborton** [comment]
I say we truncate. We might want to truncate the middle part so the file
extension is always visible too.

**Comment by mananjadhav** [comment]
> We might want to truncate the middle part so the file

Well, I thought about this. But as I mentioned earlier, we're reusing the `Header` component and the subtitle might not necessarily be a filename (have an extension). So are we sure to truncate the middle part? or you'd want to keep it multiline. We can have an assumption that the subtitle won't be that big enough but can't be sure for mobile devices.

I was thinking more on the lines of having a `subtitleTruncateMode (none, middle, tail)` in the `Header`. Would that be overkill?

**Comment by shawnborton** [comment]
Good points. I don’t think the subheader should be more than 1 line since
we don’t want to modify the height.

Perhaps we can pass through an optional prop to truncate in the middle?
Curious what Marc thinks there.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Julesssss (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by mallenexpensify** [comment]
@kbecciv I see in the OP  `Issue found by executing PR:` [#6850](https://github.com/Expensify/App/pull/6850) but I don't see a specific comment there. Do you have more context on why this issue got created?  Was it from this [convo in #expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q/p1639606010290300?thread_ts=1639445506.216300&cid=C01GTK53T8Q)?  

Upwork job posted https://www.upwork.com/jobs/~01dec90796e410ac17

**Comment by mananjadhav** [comment]
@mallenexpensify The linked PR we added the filename handling to the existing modal and it turns out that if the file name is very long it doesn't render properly. Hence this issue is created, but apparently, the issue exists in the current `Header` component itself. 

> It sounds like you are saying that anything with a "long subtitle" and using the Header component would cause this behavior and in that case probably does not need to be fixed before we can close out the linked issue.

Hence better if we treat this as a separate issue. I can submit a proposal based on Shawn's comment if we agree.

**Comment by marcaaron** [comment]
Yes, definitely agree it's a separate issue.

**Comment by mananjadhav** [comment]
**Proposal**

In the `Header` component,

1. I'll implement a prop `hasFileNameInSubtitle`, which would either be Boolean.

2. If the prop is set then I'll implement what we've done in `ReportTypingIndicator`. Two Views split by `.`. The first `View` and `Text` will have ellipsis with `flex-shrink: 1`. The second `View` with `flex-shrink: 0`, which shall have the extension.

Quick highlights,
1. Based on https://github.com/Expensify/App/issues/6873#issuecomment-999886023,  we're sure we won't have multi-line subtitles
2. Because ellipsisMode="middle" only works on iOS and Android, I've replicated what we've done in `ReportingTypingIndicator`. Hence I've added the prop, `hasFileNameInSubtitle` and not a generic prop `shouldShowEllipsis`.

**Comment by parasharrajat** [comment]
I see that you are going to borrow a lot of logic from `ReportingTypingIndicator`. It indicates that we need to dry this up. I would suggest creating a small component that does that for us and reusing it.

> I'll implement a prop hasFileNameInSubtitle, which would either be Boolean.

The header is supposed to be generic and `hasFileNameInSubtitle` sounds like a very specific case. so I have a second thought. 
Why don't we pass a component in the subtitle prop? Thus subtitle would accept a component or string.
@Julesssss  What do you say?

**Comment by mananjadhav** [comment]
> Why don't we pass a component in the subtitle prop? Thus subtitle would accept a component or string.

Like that idea @parasharrajat. Yeah I can create a new component that replicates the logic and replace at both the places.

**Comment by parasharrajat** [comment]
@mananjadhav Would you like to revise your proposal?

**Comment by mananjadhav** [comment]
Hi @parasharrajat, here you go.

**Updated Proposal**

1. I'll create a component `TextWithEllipses` (or suggest a better name) which shall take two props `leadingText` and  `trailingText`. 
2. It'll have two `Views` split by `...`. The first `View` and `Text` will have ellipsis with `flex-shrink: 1`. The second `View` with `flex-shrink: 0`.
3. In `HeaderWithCloseButton`, we'll add the prop as `subtitle: oneOf(PropTypes.func, PropTypes.string)`. Based on the type we'll either show the passed string or load the component.
4. In `AttachmentModal`, we'll add a function to render `TextWithEllipses` component by splitting the text with `.`

**Comment by parasharrajat** [comment]
Sounds good. We can look into the naming during the PR review.

@mananjadhav 's proposal looks good. 

cc: @Julesssss 
:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by marcaaron** [comment]
I don't quite understand the solution. Why are we specifying strings parts to do the truncating? Shouldn't it be determined by the length of the passed string and width of the window?

**Comment by marcaaron** [comment]
https://reactnative.dev/docs/text#ellipsizemode

**Comment by mananjadhav** [comment]
The problem with `ellipsizeMode` is that it doesn't work on Web, Desktop, and Mobile Web. I proposed a solution that we already are using in another component.

**Comment by marcaaron** [comment]
Oh right, because middle truncation is what we want to do. Ok, looks like people have been debating whether to add this to the CSS standard for a long time.

**Comment by mananjadhav** [comment]
> Oh right, because middle truncation is what we want to do.

That's right. Are you fine with the proposed solution https://github.com/Expensify/App/issues/6873#issuecomment-1004158725 ?

**Comment by marcaaron** [comment]
Gonna leave this one for @Julesssss

**Comment by Julesssss** [comment]
Caught up with the conversations and the proposal makes sense.

**Comment by MelvinBot** [comment]
📣 @mananjadhav You have been assigned to this job by @Julesssss!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by mallenexpensify** [comment]
@mananjadhav can you please apply here and I'll hire ya? 
https://www.upwork.com/jobs/~01dec90796e410ac17

**Comment by mananjadhav** [comment]
Done @mallenexpensify

**Comment by mallenexpensify** [comment]
@parasharrajat and @mananjadhav are  making progress on the PR https://github.com/Expensify/App/pull/7097

**Comment by mananjadhav** [comment]
PR is complete. There's one issue with react-native-web for which Rajat has raised a PR. Once merged we'll merge this (might have to rollback my extra commits)

**Comment by parasharrajat** [comment]
Depency PR is merged today. @mananjadhav you can proceed.

**Comment by mananjadhav** [comment]
Thanks @parasharrajat I'll push the revert today.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.32-0 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/7097


If no regressions arise, payment will be issued on 2022-02-01. :confetti_ball:

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by mananjadhav** [comment]
@mallenexpensify Reminder for Upwork.

**Comment by mallenexpensify** [comment]
Thanks for the ping @mananjadhav , you're paid now. 
@parasharrajat , I hired you, let me know when you accept and I'll pay for C+

**Comment by mallenexpensify** [comment]
Paid @parasharrajat !
